sql/src/main/java/io/crate/execution/engine/join/JoinOperations.java

```diff
@@ -113,4 +125,93 @@ public static Projection createJoinProjection(List<Symbol> outputs, List<Symbol>
             new InputColumns.SourceSymbols(joinOutputs));
         return new EvalProjection(projectionOutputs);
     }
+
+
+    public static LinkedHashMap<Set<QualifiedName>, JoinPair> buildRelationsToJoinPairsMap(List<JoinPair> joinPairs) {
+        LinkedHashMap<Set<QualifiedName>, JoinPair> joinPairsMap = new LinkedHashMap<>();
+        for (JoinPair joinPair : joinPairs) {
+            if (joinPair.condition() == null) {
+                continue;
+            }
+            JoinPair prevPair = joinPairsMap.put(Sets.newHashSet(joinPair.left(), joinPair.right()), joinPair);
+            if (prevPair != null) {
+                throw new IllegalStateException("joinPairs contains duplicate: " + joinPair + " matches " + prevPair);
+            }
+        }
+        return joinPairsMap;
+    }
+
+    /**
+     * Converts any implicit join conditions of the WHERE clause to explicit {@link JoinPair}.
+     * Every join condition that gets to be converted is removed from the {@code splitQueries}
+     * @param explicitJoinPairs The explicitJoinPairs as originally written in the query
+     * @param splitQueries      The remaining queries of the WHERE clause split by involved relations
+     * @return the new list of {@link JoinPair}
+     */
+    public static List<JoinPair> convertImplicitJoinConditionsToJoinPairs(List<JoinPair> explicitJoinPairs,
+                                                                          Map<Set<QualifiedName>, Symbol> splitQueries) {
+        LongBitSet processedJoinPairs = new LongBitSet(explicitJoinPairs.size());
```

Comment:
Wouldn't a regular BitSet be enough? We're bound by `Integer.MAX_SIZE` due to the collection limits anyway.